adoc/chapters/architecture.adoc

```diff
@@ -1099,23 +1098,33 @@ to the buffer or USM allocation from device code.
 
 ==== Forward progress
 
-Memory consistency guarantees are independent of any forward progress
-guarantees.  A SYCL implementation must execute work-items concurrently
-and must ensure that the work-items in a group obey the semantics of
-<<group-barrier,group barriers>>, but are not required to provide any
-additional forward progress guarantees.
+This section, and any subsequent section referring to progress guarantees, uses
+the following terms as defined in the {cpp} core language: thread of
+execution; weakly parallel forward progress guarantees; parallel forward
+progress guarantees; concurrent forward progress guarantees; and block with
+forward progress guarantee delegation.
 
-Synchronizing work-items via memory operations is unsafe in general, but is
-supported if and only if the following conditions are true:
+Each work-item in SYCL is a separate thread of execution, providing at least
+weakly parallel forward progress guarantees.  Whether work-items provide
+stronger forward progress guarantees is implementation-defined.
 
-  * acquire-release or sequentially consistent memory ordering is supported
-    at the scope containing the set of work-items being synchronized;
+All implementations must additionally ensure that the work-items in a group
+obey the semantics of <<group-barrier,group barriers>>: when a work-item
+arrives at a group barrier acting on group _G_, implementations must eventually
+select and potentially strengthen another work-item in group _G_ that has not
```

Comment:
```suggestion
select and potentially strengthen the progress guarantees of another work-item in group _G_ that has not
```

Reply:
My motivation for this comment is SYCL programmers might not be experts in the C++ forward progress, but we should try to give as many helpful hints here to make it easy to read the spec, write SYCL, and dive into C++ later if they want.

Reply:
I agree with the comment about readability, but I'm less sure about the fix.

I modelled this on the language used by the C++ draft, which always talks about strengthening a thread of execution, rather than strengthening a thread of execution's progress guarantees.  I'd like to try and avoid introducing minor differences in wording if we can, just to be safe.

If we rewrote this paragraph to explain the semantics of group barriers in words, would it be clear enough?  Something like:

> All implementations must additionally ensure that a work-item arriving at a
> <<group-barrier, group barrier>> does not prevent other work-items in the same
> group from making progress. When a work-item arrives at a group barrier acting
> on group _G_, implementations must eventually select and potentially strengthen
> another work-item in group _G_ that has not yet arrived at the barrier.

Reply:
Fair enough - better to follow the C++ wording. The new paragraph looks fine to me too.

Reply:
Sorry, it's just I've been burnt enough times by unintentionally misusing words like "synchronize" that it's made me paranoid. 😆 

I've put the new paragraph in 386749ce9ce4b372d7e574647c78955a1dd01d10.  Thanks again for pointing out the readability/accessibility issue here.